olive/passes/onnx/perf_tuning.py

```diff
@@ -163,99 +169,153 @@ def threads_num_tuning(model, data_root, latency_metric, config, tuning_combo):
     import psutil
 
-    if test_params["session_options"].get("extra_session_config"):
-        extra_session_config = test_params["session_options"].get("extra_session_config")
-        if extra_session_config.get("session.intra_op_thread_affinities"):
-            affinity_str = extra_session_config.get("session.intra_op_thread_affinities")
+    # prepare the inter_op_num_threads/intra_op_num_threads to be tune.
+    extra_session_config = test_params["session_options"].get("extra_session_config")
+    if extra_session_config:
+        affinity_str = extra_session_config.get("session.intra_op_thread_affinities")
+        if affinity_str:
             test_params["session_options"]["intra_op_num_threads"] = get_thread_affinity_nums(affinity_str) + 1
             threads_names = ["inter_op_num_threads"]
     elif test_params["session_options"].get("execution_mode") == ort.ExecutionMode.ORT_SEQUENTIAL:
         threads_names = ["intra_op_num_threads"]
     else:
         threads_names = ["inter_op_num_threads", "intra_op_num_threads"]
-    best_latency = None
 
-    if test_params["session_options"].get("inter_op_num_threads") and test_params["session_options"].get(
-        "intra_op_num_threads"
+    tuning_results = []
+
+    if (
+        test_params["session_options"].get("inter_op_num_threads") is not None
+        and test_params["session_options"].get("intra_op_num_threads") is not None
     ):
-        test_result = get_benchmark(model, data_root, latency_metric, config, test_params)
+        # If user specify both inter_op_num_threads and intra_op_num_threads, we will not do tuning
+        test_result = get_benchmark(model, data_root, latency_metric, config, test_params, io_bind)
         tuning_results.append(test_result)
-        best_latency = test_result["latency_ms"]
-    else:
-        for threads_name in threads_names:
-            thread_num = test_params["session_options"].get(threads_name)
-            if thread_num:
-                upper_threads_num = thread_num
-                lower_threads_num = thread_num
+        return tuning_results
+
+    for threads_name in threads_names:
+        # set the upper bound and lower bound for binary search
+        thread_num = test_params["session_options"].get(threads_name)
+        if thread_num is not None:
+            upper_threads_num = thread_num
+            lower_threads_num = thread_num
+        else:
+            upper_threads_num = config.cpu_cores or psutil.cpu_count(logical=False)
+            lower_threads_num = 1
+
+        current_threads_num = lower_threads_num
+        best_latency = None
+        best_threads_num = None
+
+        while lower_threads_num < upper_threads_num:
+            test_params["session_options"][threads_name] = current_threads_num
+            test_result = get_benchmark(model, data_root, latency_metric, config, test_params, io_bind)
+            tuning_results.append(test_result)
+
+            if best_latency is None:
```

Comment:
How about to evaluate the lower and upper point beyond the for-loop? Logics can get cleaner.

Reply:
```python

def eva(x):
    return x*x


l = -10
r = 10

l_rls, r_rls = eva(l), eva(r)

best_rls, best_idx = (l_rls, l) if l_rls < r_rls else (r_rls, r)

while l < r:
    mid = (l+r) // 2
    mid_rls = eva(mid)
    if mid < best_rls:
        # mid < best_rls, mid is better
        if best_idx == r:
            l = mid+1
        else:
            r = mid
        best_rls = mid_rls
        best_idx = mid
    else:
        # mid > best_rls, best_rls is better
        if best_idx == r:
            r = mid
        else:
            l = mid+1
print(best_rls)

```

Reply:
I'd like keep current logic to avoid calling the get_benchmark manually. 
if the refactor could remove the check in the while loop, I will do it. However, I could not remove the index check. So I will keep current logic.